New module API event for tracking authentication attempts

rjd15372 opened the pull request:
In this commit we introduce a new module API event called `ValkeyModuleEvent_AuthenticationAttempt` to track successful/failed authentication attempts.

This event will fill a struct, called `ValkeyModuleAuthenticationInfo`, with the client ID of the connection, the username, the module name  that handle the authentication event, and the result of the authentication attempt.

Fixes: #2211

rjd15372 committed:
New module API event for tracking authentication attempts

In this commit we introduce a new module API event called
`ValkeyModuleEvent_AuthenticationAttempt` to track successful/failed
authentication attempts.

This event will fill a struct, called `ValkeyModuleAuthenticationInfo`,
with the client ID of the connection, the IP address and port, the
username, the module name  that handle the authentication event, and
the result of the authentication attempt.

Fixes: #2211

Signed-off-by: Ricardo Dias <ricardo.dias@percona.com>

rjd15372 committed:
Fix clang-format errors

Signed-off-by: Ricardo Dias <ricardo.dias@percona.com>

hpatro commented on src/acl.c:
```diff
@@ -1453,16 +1454,32 @@ void addAuthErrReply(client *c, robj *err) {
+    }
+
+    {
+        ValkeyModuleAuthenticationInfo info = VALKEYMODULE_AUTHENTICATIONINFO_INITIALIZER_V1;
+        info.client_id = c->id;
+        int port;
+        connAddrPeerName(c->conn, info.addr, sizeof(info.addr), &port);
+        info.port = port;
+        info.username = username->ptr;
+        info.module_name = NULL;
+        info.success = (result == AUTH_OK);
+        moduleFireServerEvent(VALKEYMODULE_EVENT_AUTHENTICATION_ATTEMPT, 0, &info);
```
is `AUTH_OK` or `AUTH_ERR` enough for a module or should we have reason also passed down to the module?

rjd15372 replied:
I don't have a strong opinion, but since we also don't log a reason in the ACL log, I followed the same here.

hpatro commented on src/valkeymodule.h:
```diff
@@ -780,6 +782,21 @@ typedef struct ValkeyModuleKeyInfo {
+typedef struct ValkeyModuleAuthenticationInfo {
+    uint64_t version;        /* Version of this structure for ABI compat. */
+    uint64_t client_id;      /* Client ID. */
+    char addr[46];           /* IPv4 or IPv6 address. */
```
```
    char addr[NET_IP_STR_LEN];   /* IPv4 or IPv6 address. */
```

rjd15372 replied:
`NET_IP_STR_LEN` is defined in `connection.h`, which cannot be included in this file. So either we keep the `46`, or copy the macro definition to this file. What do you prefer?

hpatro commented on tests/modules/hooks.c:
```diff
@@ -205,7 +205,7 @@ void persistenceCallback(ValkeyModuleCtx *ctx, ValkeyModuleEvent e, uint64_t sub
     ValkeyModule_Log(ctx, "warning", "module-event-%s", keyname);
     if (sub == VALKEYMODULE_SUBEVENT_PERSISTENCE_SYNC_RDB_START ||
-        sub == VALKEYMODULE_SUBEVENT_PERSISTENCE_SYNC_AOF_START) 
+        sub == VALKEYMODULE_SUBEVENT_PERSISTENCE_SYNC_AOF_START)
```
remove

hpatro commented on tests/unit/moduleapi/hooks.tcl:
```diff
@@ -26,7 +27,7 @@ tags "modules" {
             assert_equal [r ping] PONG
             $rd close
-        } 
+        }
```
remove

hpatro commented on tests/unit/moduleapi/hooks.tcl:
```diff
@@ -298,7 +339,7 @@ tags "modules" {
 
         test {Test configchange hooks} {
-            r config set rdbcompression no 
+            r config set rdbcompression no
```
remove

rjd15372 committed:
Remove unecessary changes

Signed-off-by: Ricardo Dias <ricardo.dias@percona.com>

murphyjacob4 commented on src/valkeymodule.h:
```diff
@@ -780,6 +782,21 @@ typedef struct ValkeyModuleKeyInfo {
+    uint64_t version;        /* Version of this structure for ABI compat. */
+    uint64_t client_id;      /* Client ID. */
+    char addr[46];           /* IPv4 or IPv6 address. */
+    uint16_t port;           /* TCP port. */
```
addr and port are accessible through ValkeyModule_GetClientInfoById. In my opinion - we should keep the struct minimal and make the module use this API to get more details. WDYT?

rjd15372 replied:
I agree with you.

murphyjacob4 commented on src/module.c:
```diff
@@ -8067,17 +8067,41 @@ int checkModuleAuthentication(client *c, robj *username, robj *password, robj **
         serverAssert(result == VALKEYMODULE_AUTH_HANDLED);
         return AUTH_BLOCKED;
```
Do we also need to trigger this in VM_AuthenticateClientWithACLUser? For asynchronous authentication?

rjd15372 replied:
No need. When the client unblocks it will go through this code path again and will call the callback below.

murphyjacob4 commented on src/valkeymodule.h:
```diff
@@ -780,6 +782,21 @@ typedef struct ValkeyModuleKeyInfo {
+    uint16_t port;           /* TCP port. */
+    const char *username;    /* Username used for authentication. */
+    const char *module_name; /* Name of the module that is handling the authentication. */
+    int success;             /* Result of the authentication, 1 for success, 0 for failure. */
```
Not a strong opinion - but does an enum make more sense for future proofing? If so, maybe naming it GRANTED/DENIED makes more sense, so we can differentiate a successful denial from an internal issue (timeout, etc).

hpatro replied:
I like this suggestion. Kind of boils down to my request where we're able to differentiate about the failure.

rjd15372 replied:
Me too. I'll use an enum.

rjd15372 replied:
Unfortunately we can't distinguish between a password mismatch or an authentication failure because the module API to handle authentication does not provide that information.

rjd15372 committed:
Addresses reviewers suggestions

- Used an enumeration for the authentication result
- removed IP and port from auth info structure
- added tests to test the module blocking authentication

Signed-off-by: Ricardo Dias <ricardo.dias@percona.com>

madolson commented on src/module.c:
```diff
@@ -11763,6 +11786,23 @@ static uint64_t moduleEventVersions[] = {
  *
  *         ValkeyModuleKey *key;    // Key name
  *
+ * * ValkeyModule_Authentication_Attempt
```
It looks like you modified the struct since you wrote the docs. Looking up the clients looks like the right solution to extract the ip.

madolson commented on src/module.c:
```diff
@@ -8067,17 +8067,39 @@ int checkModuleAuthentication(client *c, robj *username, robj *password, robj **
         return AUTH_BLOCKED;
     }
+
+    ValkeyModuleAuthCtx *auth_ctx = c->module_data ? c->module_data->module_auth_ctx : NULL;
```
I assume we always want c->module_data to be set, do we know what cases it's null now?

rjd15372 replied:
`c->module_data` will be null if no module has registered an authentication callback. This corresponds to the common code path when there are no auth modules loaded in the server.

rjd15372 commented:
@madolson can we get this in for 9.0?

rjd15372 committed:
Fix code comment

Signed-off-by: Ricardo Dias <ricardo.dias@percona.com>

rjd15372 committed:
Merge remote-tracking branch 'upstream/unstable' into auth-callback-api

enjoy-binbin reviewed (commented):
LGTM overall.

zuiderkwast commented on src/module.c:
```diff
@@ -11763,6 +11786,20 @@ static uint64_t moduleEventVersions[] = {
  *
  *         ValkeyModuleKey *key;    // Key name
  *
+ * * ValkeyModule_Authentication_Attempt
```
This seems wrong but I'm not sure which struct or value is referenced here. (Maybe other event names in this comment are wrong too?)

```suggestion
 * * ValkeyModuleEvent_AuthenticationAttempt
```

madolson replied:
I think it's supposed to be the define enum, `VALKEYMODULE_EVENT_AUTHENTICATION_ATTEMPT`. I'm not sure why these aren't in all CAPs.

zuiderkwast replied:
We should fix it...

rjd15372 replied:
@zuiderkwast should I fix the whole comment, including the above event names?

rjd15372 replied:
I fixes the function comment. The names there match the list of `ValkeyModuleEvent` variables declared in valkeymodule.h

rjd15372 committed:
Fix `VM_SubscribeToServerEvent` function comment

Signed-off-by: Ricardo Dias <ricardo.dias@percona.com>

zuiderkwast commented on src/acl.c:
```diff
@@ -1458,16 +1459,30 @@ void addAuthErrReply(client *c, robj *err) {
+        result = AUTH_ERR;
+    }
+
+    {
+        ValkeyModuleAuthenticationInfo info = VALKEYMODULE_AUTHENTICATIONINFO_INITIALIZER_V1;
+        info.client_id = c->id;
+        info.username = username->ptr;
+        info.module_name = NULL;
+        info.result = result == AUTH_OK ? VALKEYMODULE_AUTH_RESULT_GRANTED
+                                        : VALKEYMODULE_AUTH_RESULT_DENIED;
+        moduleFireServerEvent(VALKEYMODULE_EVENT_AUTHENTICATION_ATTEMPT, 0, &info);
```
This code is not large, but it's duplicated. If we change V1 to V2 in the future, we'll need to remember to change in both places.

Let's move this and the duplicated code in `checkModuleAuthentication` to a new function in `module.{c,h}`? e.g. `moduleFireAuthenticationEvent(client_id, username, module_name, is_granted)`.

Then, acl.c doesn't need to include valkeymodule.h. It's enough to include module.h.

rjd15372 replied:
Done. The code looks much better now. Thanks for the suggestion.

rjd15372 committed:
Refactor to remove duplicate code

Signed-off-by: Ricardo Dias <ricardo.dias@percona.com>

zuiderkwast commented on src/module.c:
```diff
@@ -11763,6 +11791,20 @@ static uint64_t moduleEventVersions[] = {
+ *         const char *module_name; // Name of the module that is handling the
+ *                                  // authentication. It is NULL if the
+ *                                  // authentication is handled by the core.
+ *         ValkeyModuleAuthenticationResult result;   // Result of the authentication.
```
This documentation is so complete that it can be used without looking at the structs in valkeymodule.h, except for the result values. Let's add them too.

```suggestion
 *         ValkeyModuleAuthenticationResult result;
 *                                  // Result of the authentication:
 *                                  // VALKEYMODULE_AUTH_RESULT_GRANTED or
 *                                  // VALKEYMODULE_AUTH_RESULT_DENIED.
```

rjd15372 committed:
Improve comment

Signed-off-by: Ricardo Dias <ricardo.dias@percona.com>
